bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/GenerateGroupByKeyDoFn.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.beam.validation;
+
+import com.google.cloud.bigtable.beam.validation.HadoopHashTableSource.RangeHash;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.values.KV;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.hbase.util.Bytes;
+
+/**
+ * A {@link DoFn} function that takes a range and hash from HBase and generates a batch or work. The
+ * key is the start key of the batch and value is the sorted range of RangeHashes. This function
+ * generates a KV<String, List<RangeHash>> to preserve the sorted nature of RangeHashes. There is no
+ * guarantee of preserving the order of RangeHash if this function returns KV<String,RangeHash>.
+ *
+ * <p>Sorted list of RangeHashes allows the subsequent functions to issue a single scan to Bigtable
+ * covering all the ranges in the list of RangeHash.
+ */
+public class GenerateGroupByKeyDoFn extends DoFn<RangeHash, KV<String, List<RangeHash>>> {
```

Comment:
> have your source emit KV<partition_start_key, RangeHash>
We can't do that since we want to maintain the order of RangeHash. Beam does not guarantee the relative ordering of elements after groupby. I have tested this. 

> use GroupIntoBatches to create smaller bundles
GroupInToBatches requires a batch key to start with. We don't have that, so even with GroupInToBatches we need to iterate over the bundle and generate the keys. I prefer to keep it simple and group linearly (we will never have multiple keys to bundle).